src/box/relay.cc

```diff
@@ -453,33 +492,50 @@ relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,
 		relay_delete(relay);
 	});
 
-	/* Freeze a read view in engines. */
+	struct synchro_request req;
+	struct raft_request raft_req;
+	struct vclock limbo_vclock;
 	struct engine_join_ctx ctx;
-	engine_prepare_join_xc(&ctx);
+	ctx.array = NULL;
+	/*
+	 * Guard is defined prematurely, as it must exist in the global scope
+	 * of the function. It works only during read-view join.
+	 */
 	auto join_guard = make_scoped_guard([&] {
 		engine_complete_join(&ctx);
 	});
+	if (cursor != NULL && cursor->is_checkpoint_join) {
+		join_guard.is_active = false;
+		/** Checkpoint initial join is going to be done. */
+		int rc = relay_prepare_checkpoint_initial_join(cursor, vclock,
+							       &raft_req, &req);
+		if (rc != 0)
+			diag_raise();
+		say_info("sending checkpoint with vclock %s at signature %"
+			 PRId64 " to replica at %s", vclock_to_string(vclock),
+			 vclock_sum(vclock), sio_socketname(io->fd));
+	} else {
+		/* Freeze a read view in engines. */
+		engine_prepare_join_xc(&ctx);
 
-	/*
-	 * Sync WAL to make sure that all changes visible from
-	 * the frozen read view are successfully committed and
-	 * obtain corresponding vclock.
-	 */
-	if (wal_sync(vclock) != 0)
-		diag_raise();
+		/*
+		 * Sync WAL to make sure that all changes visible from
+		 * the frozen read view are successfully committed and
+		 * obtain corresponding vclock.
+		 */
+		if (wal_sync(vclock) != 0)
+			diag_raise();
 
-	/*
-	 * Start sending data only when the latest sync
-	 * transaction is confirmed.
-	 */
-	if (txn_limbo_wait_confirm(&txn_limbo) != 0)
-		diag_raise();
+		/*
+		 * Start sending data only when the latest sync
+		 * transaction is confirmed.
+		 */
+		if (txn_limbo_wait_confirm(&txn_limbo) != 0)
+			diag_raise();
 
-	struct synchro_request req;
-	struct raft_request raft_req;
-	struct vclock limbo_vclock;
-	txn_limbo_checkpoint(&txn_limbo, &req, &limbo_vclock);
-	box_raft_checkpoint_local(&raft_req);
+		txn_limbo_checkpoint(&txn_limbo, &req, &limbo_vclock);
+		box_raft_checkpoint_local(&raft_req);
```

Comment:
Do we really need **not** to do it in case of checkpoint join? The code would be simpler if we did it unconditionally for both types of join.

Reply:
From Vladislav:

> The problem is that you are dumping the latest state, while the checkpoint might be old and have an outdated state. For example, the checkpoint might be from term 10, while the latest term is 100. Then how will the subscribe work after such join, when the replica will be receiving stuff from "outdated" terms?

And I agree with him, we can **not** send the latest state we have. In such case replica will have term 100 with data from the 10th term. After that during subscribe it'll get term 11 and this will be lowering term from replica's point of view, which cannot happen at all. Let's **not** do that just for code simplicity, we must recover states from the snap, we're sending

Reply:
> What do you think about moving the code sending the RAFT state to memtx_engine_join? The memtx engine is already aware about RAFT so it should be fine IMO. This would make the relay code handle checkpoint join the same way as read view join, without confusing conditions.

Let's do that

Reply:
> What do you think about moving the code sending the RAFT state to memtx_engine_join? 

@locker, maybe `engine_prepare_join` and `engine_join` are the better places for that? IMHO, sending vclock and `META_JOIN` stage are the common parts for all engine and we should not put them right into memtx, even though we know, that memtx is send firstly during join. It's completely not memtx related.

Reply:
`engine_prepare_join` and `engine_join` (as well as other `engine_*` methods) are supposed to be thin wrappers around engine callbacks. I think that sending the RAFT state in the memtx engine callback is fine because memtx is special: it stores system spaces and system data. Sending the RAFT state from `memtx_engine_join` would only be consistent with `memtx_engine_checkpoint`, which stores the RAFT state in the snapshot.